File: include/libqpicture.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const qpaintdevice_enums = @import("libqpaintdevice.zig").enums;
const std = @import("std");

/// https://doc.qt.io/qt-6/qpicture.html
pub const qpicture = struct {
    /// New constructs a new QPicture object.
    ///
    ///
    pub fn New() QtC.QPicture {
        return qtc.QPicture_new();
    }

    /// New2 constructs a new QPicture object.
    ///
    /// ``` param1: QtC.QPicture ```
    pub fn New2(param1: ?*anyopaque) QtC.QPicture {
        return qtc.QPicture_new2(@ptrCast(param1));
    }

    /// New3 constructs a new QPicture object.
    ///
    /// ``` formatVersion: i32 ```
    pub fn New3(formatVersion: i32) QtC.QPicture {
        return qtc.QPicture_new3(@intCast(formatVersion));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpicture.html#isNull)
    ///
    /// ``` self: QtC.QPicture ```
    pub fn IsNull(self: ?*anyopaque) bool {
        return qtc.QPicture_IsNull(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpicture.html#devType)
    ///
    /// ``` self: QtC.QPicture ```
    pub fn DevType(self: ?*anyopaque) i32 {
        return qtc.QPicture_DevType(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpicture.html#devType)
    ///
    /// Allows for overriding the related default method
    ///
    /// ``` self: QtC.QPicture, slot: fn () callconv(.c) i32 ```
    pub fn OnDevType(self: ?*anyopaque, slot: fn () callconv(.c) i32) void {
        qtc.QPicture_OnDevType(@ptrCast(self), @as(isize, @bitCast(@intFromPtr(&slot))));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpicture.html#devType)
    ///
    /// Base class method implementation
    ///
    /// ``` self: QtC.QPicture ```
    pub fn QBaseDevType(self: ?*anyopaque) i32 {
        return qtc.QPicture_QBaseDevType(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpicture.html#size)
    ///
    /// ``` self: QtC.QPicture ```
    pub fn Size(self: ?*anyopaque) u32 {
        return qtc.QPicture_Size(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpicture.html#data)
    ///
    /// ``` self: QtC.QPicture ```
    pub fn Data(self: ?*anyopaque) []const u8 {
        const _ret = qtc.QPicture_Data(@ptrCast(self));
        return std.mem.span(_ret);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpicture.html#setData)
    ///
    /// ``` self: QtC.QPicture, data: []const u8, size: u32 ```
    pub fn SetData(self: ?*anyopaque, data: []const u8, size: u32) void {
        const data_Cstring = @constCast(data.ptr);
        qtc.QPicture_SetData(@ptrCast(self), data_Cstring, @intCast(size));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpicture.html#setData)
    ///
    /// Allows for overriding the related default method
    ///
    /// ``` self: QtC.QPicture, slot: fn (self: QtC.QPicture, data: []const u8, size: u32) callconv(.c) void ```
    pub fn OnSetData(self: ?*anyopaque, slot: fn (?*anyopaque, []const u8, u32) callconv(.c) void) void {
        qtc.QPicture_OnSetData(@ptrCast(self), @as(isize, @bitCast(@intFromPtr(&slot))));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpicture.html#setData)
    ///
    /// Base class method implementation
    ///
    /// ``` self: QtC.QPicture, data: []const u8, size: u32 ```
    pub fn QBaseSetData(self: ?*anyopaque, data: []const u8, size: u32) void {
        const data_Cstring = @constCast(data.ptr);
        qtc.QPicture_QBaseSetData(@ptrCast(self), data_Cstring, @intCast(size));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpicture.html#play)
    ///
    /// ``` self: QtC.QPicture, p: QtC.QPainter ```
    pub fn Play(self: ?*anyopaque, p: ?*anyopaque) bool {
        return qtc.QPicture_Play(@ptrCast(self), @ptrCast(p));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpicture.html#load)
    ///
    /// ``` self: QtC.QPicture, dev: QtC.QIODevice ```
    pub fn Load(self: ?*anyopaque, dev: ?*anyopaque) bool {
        return qtc.QPicture_Load(@ptrCast(self), @ptrCast(dev));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpicture.html#load)
    ///
    /// ``` self: QtC.QPicture, fileName: []const u8 ```
    pub fn LoadWithFileName(self: ?*anyopaque, fileName: []const u8) bool {
        const fileName_str = qtc.struct_libqt_string{
            .len = fileName.len,
            .data = @constCast(fileName.ptr),
        };
        return qtc.QPicture_LoadWithFileName(@ptrCast(self), fileName_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpicture.html#save)
    ///
    /// ``` self: QtC.QPicture, dev: QtC.QIODevice ```
    pub fn Save(self: ?*anyopaque, dev: ?*anyopaque) bool {
        return qtc.QPicture_Save(@ptrCast(self), @ptrCast(dev));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpicture.html#save)
    ///
    /// ``` self: QtC.QPicture, fileName: []const u8 ```
    pub fn SaveWithFileName(self: ?*anyopaque, fileName: []const u8) bool {
        const fileName_str = qtc.struct_libqt_string{
            .len = fileName.len,
            .data = @constCast(fileName.ptr),
        };
        return qtc.QPicture_SaveWithFileName(@ptrCast(self), fileName_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpicture.html#boundingRect)
    ///
    /// ``` self: QtC.QPicture ```
    pub fn BoundingRect(self: ?*anyopaque) QtC.QRect {
        return qtc.QPicture_BoundingRect(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpicture.html#setBoundingRect)
    ///
    /// ``` self: QtC.QPicture, r: QtC.QRect ```
    pub fn SetBoundingRect(self: ?*anyopaque, r: ?*anyopaque) void {
        qtc.QPicture_SetBoundingRect(@ptrCast(self), @ptrCast(r));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpicture.html#operator=)
    ///
    /// ``` self: QtC.QPicture, p: QtC.QPicture ```
    pub fn OperatorAssign(self: ?*anyopaque, p: ?*anyopaque) void {
        qtc.QPicture_OperatorAssign(@ptrCast(self), @ptrCast(p));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpicture.html#swap)
    ///
    /// ``` self: QtC.QPicture, other: QtC.QPicture ```
    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QPicture_Swap(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpicture.html#detach)
    ///
    /// ``` self: QtC.QPicture ```
    pub fn Detach(self: ?*anyopaque) void {
        qtc.QPicture_Detach(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpicture.html#isDetached)
    ///
    /// ``` self: QtC.QPicture ```
    pub fn IsDetached(self: ?*anyopaque) bool {
        return qtc.QPicture_IsDetached(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpicture.html#paintEngine)
    ///
    /// ``` self: QtC.QPicture ```
    pub fn PaintEngine(self: ?*anyopaque) QtC.QPaintEngine {
        return qtc.QPicture_PaintEngine(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpicture.html#paintEngine)
    ///
    /// Allows for overriding the related default method
    ///
    /// ``` self: QtC.QPicture, slot: fn () callconv(.c) QtC.QPaintEngine ```
    pub fn OnPaintEngine(self: ?*anyopaque, slot: fn () callconv(.c) QtC.QPaintEngine) void {
        qtc.QPicture_OnPaintEngine(@ptrCast(self), @as(isize, @bitCast(@intFromPtr(&slot))));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpicture.html#paintEngine)
    ///
    /// Base class method implementation
    ///
    /// ``` self: QtC.QPicture ```
    pub fn QBasePaintEngine(self: ?*anyopaque) QtC.QPaintEngine {
        return qtc.QPicture_QBasePaintEngine(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpicture.html#metric)
    ///
    /// ``` self: QtC.QPicture, m: qpaintdevice_enums.PaintDeviceMetric ```
    pub fn Metric(self: ?*anyopaque, m: i64) i32 {
        return qtc.QPicture_Metric(@ptrCast(self), @intCast(m));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpicture.html#metric)
    ///
    /// Allows for overriding the related default method
    ///
    /// ``` self: QtC.QPicture, slot: fn (self: QtC.QPicture, m: qpaintdevice_enums.PaintDeviceMetric) callconv(.c) i32 ```
    pub fn OnMetric(self: ?*anyopaque, slot: fn (?*anyopaque, i64) callconv(.c) i32) void {
        qtc.QPicture_OnMetric(@ptrCast(self), @as(isize, @bitCast(@intFromPtr(&slot))));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpicture.html#metric)
    ///
    /// Base class method implementation
    ///
    /// ``` self: QtC.QPicture, m: qpaintdevice_enums.PaintDeviceMetric ```
    pub fn QBaseMetric(self: ?*anyopaque, m: i64) i32 {
        return qtc.QPicture_QBaseMetric(@ptrCast(self), @intCast(m));
    }

    /// Inherited from QPaintDevice
    ///
    /// [Qt documentation](https://doc.qt.io/qt-6/qpaintdevice.html#paintingActive)
    ///
    /// ``` self: QtC.QPicture ```
    pub fn PaintingActive(self: ?*anyopaque) bool {
        return qtc.QPaintDevice_PaintingActive(@ptrCast(self));
    }

    /// Inherited from QPaintDevice
    ///
    /// [Qt documentation](https://doc.qt.io/qt-6/qpaintdevice.html#width)
    ///
    /// ``` self: QtC.QPicture ```
    pub fn Width(self: ?*anyopaque) i32 {
        return qtc.QPaintDevice_Width(@ptrCast(self));
    }

    /// Inherited from QPaintDevice
    ///
    /// [Qt documentation](https://doc.qt.io/qt-6/qpaintdevice.html#height)
    ///
    /// ``` self: QtC.QPicture ```
    pub fn Height(self: ?*anyopaque) i32 {
        return qtc.QPaintDevice_Height(@ptrCast(self));
    }

    /// Inherited from QPaintDevice
    ///
    /// [Qt documentation](https://doc.qt.io/qt-6/qpaintdevice.html#widthMM)
    ///
    /// ``` self: QtC.QPicture ```
    pub fn WidthMM(self: ?*anyopaque) i32 {
        return qtc.QPaintDevice_WidthMM(@ptrCast(self));
    }

    /// Inherited from QPaintDevice
    ///
    /// [Qt documentation](https://doc.qt.io/qt-6/qpaintdevice.html#heightMM)
    ///
    /// ``` self: QtC.QPicture ```
    pub fn HeightMM(self: ?*anyopaque) i32 {
        return qtc.QPaintDevice_HeightMM(@ptrCast(self));
    }

    /// Inherited from QPaintDevice
    ///
    /// [Qt documentation](https://doc.qt.io/qt-6/qpaintdevice.html#logicalDpiX)
    ///
    /// ``` self: QtC.QPicture ```
    pub fn LogicalDpiX(self: ?*anyopaque) i32 {
        return qtc.QPaintDevice_LogicalDpiX(@ptrCast(self));
    }

    /// Inherited from QPaintDevice
    ///
    /// [Qt documentation](https://doc.qt.io/qt-6/qpaintdevice.html#logicalDpiY)
    ///
    /// ``` self: QtC.QPicture ```
    pub fn LogicalDpiY(self: ?*anyopaque) i32 {
        return qtc.QPaintDevice_LogicalDpiY(@ptrCast(self));
    }

    /// Inherited from QPaintDevice
    ///
    /// [Qt documentation](https://doc.qt.io/qt-6/qpaintdevice.html#physicalDpiX)
    ///
    /// ``` self: QtC.QPicture ```
    pub fn PhysicalDpiX(self: ?*anyopaque) i32 {
        return qtc.QPaintDevice_PhysicalDpiX(@ptrCast(self));
    }

    /// Inherited from QPaintDevice
    ///
    /// [Qt documentation](https://doc.qt.io/qt-6/qpaintdevice.html#physicalDpiY)
    ///
    /// ``` self: QtC.QPicture ```
    pub fn PhysicalDpiY(self: ?*anyopaque) i32 {
        return qtc.QPaintDevice_PhysicalDpiY(@ptrCast(self));
    }

    /// Inherited from QPaintDevice
    ///
    /// [Qt documentation](https://doc.qt.io/qt-6/qpaintdevice.html#devicePixelRatio)
    ///
    /// ``` self: QtC.QPicture ```
    pub fn DevicePixelRatio(self: ?*anyopaque) f64 {
        return qtc.QPaintDevice_DevicePixelRatio(@ptrCast(self));
    }

    /// Inherited from QPaintDevice
    ///
    /// [Qt documentation](https://doc.qt.io/qt-6/qpaintdevice.html#devicePixelRatioF)
    ///
    /// ``` self: QtC.QPicture ```
    pub fn DevicePixelRatioF(self: ?*anyopaque) f64 {
        return qtc.QPaintDevice_DevicePixelRatioF(@ptrCast(self));
    }

    /// Inherited from QPaintDevice
    ///
    /// [Qt documentation](https://doc.qt.io/qt-6/qpaintdevice.html#colorCount)
    ///
    /// ``` self: QtC.QPicture ```
    pub fn ColorCount(self: ?*anyopaque) i32 {
        return qtc.QPaintDevice_ColorCount(@ptrCast(self));
    }

    /// Inherited from QPaintDevice
    ///
    /// [Qt documentation](https://doc.qt.io/qt-6/qpaintdevice.html#depth)
    ///
    /// ``` self: QtC.QPicture ```
    pub fn Depth(self: ?*anyopaque) i32 {
        return qtc.QPaintDevice_Depth(@ptrCast(self));
    }

    /// Inherited from QPaintDevice
    ///
    /// [Qt documentation](https://doc.qt.io/qt-6/qpaintdevice.html#devicePixelRatioFScale)
    ///
    ///
    pub fn DevicePixelRatioFScale() f64 {
        return qtc.QPaintDevice_DevicePixelRatioFScale();
    }

    /// Inherited from QPaintDevice
    ///
    /// [Qt documentation](https://doc.qt.io/qt-6/qpaintdevice.html#initPainter)
    ///
    /// Wrapper to allow calling virtual or protected method
    ///
    /// ``` self: QtC.QPicture, painter: QtC.QPainter ```
    pub fn InitPainter(self: ?*anyopaque, painter: ?*anyopaque) void {
        qtc.QPicture_InitPainter(@ptrCast(self), @ptrCast(painter));
    }

    /// Inherited from QPaintDevice
    ///
    /// [Qt documentation](https://doc.qt.io/qt-6/qpaintdevice.html#initPainter)
    ///
    /// Wrapper to allow calling base class virtual or protected method
    ///
    /// ``` self: QtC.QPicture, painter: QtC.QPainter ```
    pub fn QBaseInitPainter(self: ?*anyopaque, painter: ?*anyopaque) void {
        qtc.QPicture_QBaseInitPainter(@ptrCast(self), @ptrCast(painter));
    }

    /// Inherited from QPaintDevice
    ///
    /// [Qt documentation](https://doc.qt.io/qt-6/qpaintdevice.html#initPainter)
    ///
    /// Wrapper to allow overriding base class virtual or protected method
    ///
    /// ``` self: QtC.QPicture, slot: fn (self: QtC.QPicture, painter: QtC.QPainter) callconv(.c) void ```
    pub fn OnInitPainter(self: ?*anyopaque, slot: fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QPicture_OnInitPainter(@ptrCast(self), @as(isize, @bitCast(@intFromPtr(&slot))));
    }

    /// Inherited from QPaintDevice
    ///
    /// [Qt documentation](https://doc.qt.io/qt-6/qpaintdevice.html#redirected)
    ///
    /// Wrapper to allow calling virtual or protected method
    ///
    /// ``` self: QtC.QPicture, offset: QtC.QPoint ```
    pub fn Redirected(self: ?*anyopaque, offset: ?*anyopaque) QtC.QPaintDevice {
        return qtc.QPicture_Redirected(@ptrCast(self), @ptrCast(offset));
    }

    /// Inherited from QPaintDevice
    ///
    /// [Qt documentation](https://doc.qt.io/qt-6/qpaintdevice.html#redirected)
    ///
    /// Wrapper to allow calling base class virtual or protected method
    ///
    /// ``` self: QtC.QPicture, offset: QtC.QPoint ```
    pub fn QBaseRedirected(self: ?*anyopaque, offset: ?*anyopaque) QtC.QPaintDevice {
        return qtc.QPicture_QBaseRedirected(@ptrCast(self), @ptrCast(offset));
    }

    /// Inherited from QPaintDevice
    ///
    /// [Qt documentation](https://doc.qt.io/qt-6/qpaintdevice.html#redirected)
    ///
    /// Wrapper to allow overriding base class virtual or protected method
    ///
    /// ``` self: QtC.QPicture, slot: fn (self: QtC.QPicture, offset: QtC.QPoint) callconv(.c) QtC.QPaintDevice ```
    pub fn OnRedirected(self: ?*anyopaque, slot: fn (?*anyopaque, ?*anyopaque) callconv(.c) QtC.QPaintDevice) void {
        qtc.QPicture_OnRedirected(@ptrCast(self), @as(isize, @bitCast(@intFromPtr(&slot))));
    }

    /// Inherited from QPaintDevice
    ///
    /// [Qt documentation](https://doc.qt.io/qt-6/qpaintdevice.html#sharedPainter)
    ///
    /// Wrapper to allow calling virtual or protected method
    ///
    /// ``` self: QtC.QPicture ```
    pub fn SharedPainter(self: ?*anyopaque) QtC.QPainter {
        return qtc.QPicture_SharedPainter(@ptrCast(self));
    }

    /// Inherited from QPaintDevice
    ///
    /// [Qt documentation](https://doc.qt.io/qt-6/qpaintdevice.html#sharedPainter)
    ///
    /// Wrapper to allow calling base class virtual or protected method
    ///
    /// ``` self: QtC.QPicture ```
    pub fn QBaseSharedPainter(self: ?*anyopaque) QtC.QPainter {
        return qtc.QPicture_QBaseSharedPainter(@ptrCast(self));
    }

    /// Inherited from QPaintDevice
    ///
    /// [Qt documentation](https://doc.qt.io/qt-6/qpaintdevice.html#sharedPainter)
    ///
    /// Wrapper to allow overriding base class virtual or protected method
    ///
    /// ``` self: QtC.QPicture, slot: fn () callconv(.c) QtC.QPainter ```
    pub fn OnSharedPainter(self: ?*anyopaque, slot: fn () callconv(.c) QtC.QPainter) void {
        qtc.QPicture_OnSharedPainter(@ptrCast(self), @as(isize, @bitCast(@intFromPtr(&slot))));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpicture.html#dtor.QPicture)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.QPicture ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QPicture_Delete(@ptrCast(self));
    }
};

File: src/libqpicture.cpp
#include <QIODevice>
#include <QPaintDevice>
#include <QPaintEngine>
#include <QPainter>
#include <QPicture>
#include <QPoint>
#include <QRect>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qpicture.h>
#include "libqpicture.h"
#include "libqpicture.hxx"

QPicture* QPicture_new() {
    return new VirtualQPicture();
}

QPicture* QPicture_new2(const QPicture* param1) {
    return new VirtualQPicture(*param1);
}

QPicture* QPicture_new3(int formatVersion) {
    return new VirtualQPicture(static_cast<int>(formatVersion));
}

bool QPicture_IsNull(const QPicture* self) {
    return self->isNull();
}

int QPicture_DevType(const QPicture* self) {
    auto* vqpicture = dynamic_cast<const VirtualQPicture*>(self);
    if (vqpicture && vqpicture->isVirtualQPicture) {
        return self->devType();
    } else {
        return ((VirtualQPicture*)self)->devType();
    }
}

// Subclass method to allow providing a virtual method re-implementation
void QPicture_OnDevType(const QPicture* self, intptr_t slot) {
    auto* vqpicture = const_cast<VirtualQPicture*>(dynamic_cast<const VirtualQPicture*>(self));
    if (vqpicture && vqpicture->isVirtualQPicture) {
        vqpicture->setQPicture_DevType_Callback(reinterpret_cast<VirtualQPicture::QPicture_DevType_Callback>(slot));
    }
}

// Virtual base class handler implementation
int QPicture_QBaseDevType(const QPicture* self) {
    auto* vqpicture = dynamic_cast<const VirtualQPicture*>(self);
    if (vqpicture && vqpicture->isVirtualQPicture) {
        vqpicture->setQPicture_DevType_IsBase(true);
        return vqpicture->devType();
    } else {
        return ((VirtualQPicture*)self)->devType();
    }
}

unsigned int QPicture_Size(const QPicture* self) {
    return static_cast<unsigned int>(self->size());
}

const char* QPicture_Data(const QPicture* self) {
    return (const char*)self->data();
}

void QPicture_SetData(QPicture* self, const char* data, unsigned int size) {
    auto* vqpicture = dynamic_cast<VirtualQPicture*>(self);
    if (vqpicture && vqpicture->isVirtualQPicture) {
        self->setData(data, static_cast<uint>(size));
    } else {
        ((VirtualQPicture*)self)->setData(data, static_cast<uint>(size));
    }
}

// Subclass method to allow providing a virtual method re-implementation
void QPicture_OnSetData(QPicture* self, intptr_t slot) {
    auto* vqpicture = dynamic_cast<VirtualQPicture*>(self);
    if (vqpicture && vqpicture->isVirtualQPicture) {
        vqpicture->setQPicture_SetData_Callback(reinterpret_cast<VirtualQPicture::QPicture_SetData_Callback>(slot));
    }
}

// Virtual base class handler implementation
void QPicture_QBaseSetData(QPicture* self, const char* data, unsigned int size) {
    auto* vqpicture = dynamic_cast<VirtualQPicture*>(self);
    if (vqpicture && vqpicture->isVirtualQPicture) {
        vqpicture->setQPicture_SetData_IsBase(true);
        vqpicture->setData(data, static_cast<uint>(size));
    } else {
        ((VirtualQPicture*)self)->setData(data, static_cast<uint>(size));
    }
}

bool QPicture_Play(QPicture* self, QPainter* p) {
    return self->play(p);
}

bool QPicture_Load(QPicture* self, QIODevice* dev) {
    return self->load(dev);
}

bool QPicture_Load2(QPicture* self, const libqt_string fileName) {
    QString fileName_QString = QString::fromUtf8(fileName.data, fileName.len);
    return self->load(fileName_QString);
}

bool QPicture_Save(QPicture* self, QIODevice* dev) {
    return self->save(dev);
}

bool QPicture_Save2(QPicture* self, const libqt_string fileName) {
    QString fileName_QString = QString::fromUtf8(fileName.data, fileName.len);
    return self->save(fileName_QString);
}

QRect* QPicture_BoundingRect(const QPicture* self) {
    return new QRect(self->boundingRect());
}

void QPicture_SetBoundingRect(QPicture* self, const QRect* r) {
    self->setBoundingRect(*r);
}

void QPicture_OperatorAssign(QPicture* self, const QPicture* p) {
    self->operator=(*p);
}

void QPicture_Swap(QPicture* self, QPicture* other) {
    self->swap(*other);
}

void QPicture_Detach(QPicture* self) {
    self->detach();
}

bool QPicture_IsDetached(const QPicture* self) {
    return self->isDetached();
}

QPaintEngine* QPicture_PaintEngine(const QPicture* self) {
    auto* vqpicture = dynamic_cast<const VirtualQPicture*>(self);
    if (vqpicture && vqpicture->isVirtualQPicture) {
        return self->paintEngine();
    } else {
        return ((VirtualQPicture*)self)->paintEngine();
    }
}

// Subclass method to allow providing a virtual method re-implementation
void QPicture_OnPaintEngine(const QPicture* self, intptr_t slot) {
    auto* vqpicture = const_cast<VirtualQPicture*>(dynamic_cast<const VirtualQPicture*>(self));
    if (vqpicture && vqpicture->isVirtualQPicture) {
        vqpicture->setQPicture_PaintEngine_Callback(reinterpret_cast<VirtualQPicture::QPicture_PaintEngine_Callback>(slot));
    }
}

// Virtual base class handler implementation
QPaintEngine* QPicture_QBasePaintEngine(const QPicture* self) {
    auto* vqpicture = dynamic_cast<const VirtualQPicture*>(self);
    if (vqpicture && vqpicture->isVirtualQPicture) {
        vqpicture->setQPicture_PaintEngine_IsBase(true);
        return vqpicture->paintEngine();
    } else {
        return ((VirtualQPicture*)self)->paintEngine();
    }
}

int QPicture_Metric(const QPicture* self, int m) {
    auto* vqpicture = dynamic_cast<const VirtualQPicture*>(self);
    if (vqpicture && vqpicture->isVirtualQPicture) {
        return vqpicture->metric(static_cast<QPaintDevice::PaintDeviceMetric>(m));
    }
    return {};
}

// Subclass method to allow providing a virtual method re-implementation
void QPicture_OnMetric(const QPicture* self, intptr_t slot) {
    auto* vqpicture = const_cast<VirtualQPicture*>(dynamic_cast<const VirtualQPicture*>(self));
    if (vqpicture && vqpicture->isVirtualQPicture) {
        vqpicture->setQPicture_Metric_Callback(reinterpret_cast<VirtualQPicture::QPicture_Metric_Callback>(slot));
    }
}

// Virtual base class handler implementation
int QPicture_QBaseMetric(const QPicture* self, int m) {
    auto* vqpicture = dynamic_cast<const VirtualQPicture*>(self);
    if (vqpicture && vqpicture->isVirtualQPicture) {
        vqpicture->setQPicture_Metric_IsBase(true);
        return vqpicture->metric(static_cast<QPaintDevice::PaintDeviceMetric>(m));
    }
    return {};
}

// Derived class handler implementation
void QPicture_InitPainter(const QPicture* self, QPainter* painter) {
    auto* vqpicture = const_cast<VirtualQPicture*>(dynamic_cast<const VirtualQPicture*>(self));
    if (vqpicture && vqpicture->isVirtualQPicture) {
        vqpicture->initPainter(painter);
    } else {
        ((VirtualQPicture*)self)->initPainter(painter);
    }
}

// Base class handler implementation
void QPicture_QBaseInitPainter(const QPicture* self, QPainter* painter) {
    auto* vqpicture = const_cast<VirtualQPicture*>(dynamic_cast<const VirtualQPicture*>(self));
    if (vqpicture && vqpicture->isVirtualQPicture) {
        vqpicture->setQPicture_InitPainter_IsBase(true);
        vqpicture->initPainter(painter);
    } else {
        ((VirtualQPicture*)self)->initPainter(painter);
    }
}

// Auxiliary method to allow providing re-implementation
void QPicture_OnInitPainter(const QPicture* self, intptr_t slot) {
    auto* vqpicture = const_cast<VirtualQPicture*>(dynamic_cast<const VirtualQPicture*>(self));
    if (vqpicture && vqpicture->isVirtualQPicture) {
        vqpicture->setQPicture_InitPainter_Callback(reinterpret_cast<VirtualQPicture::QPicture_InitPainter_Callback>(slot));
    }
}

// Derived class handler implementation
QPaintDevice* QPicture_Redirected(const QPicture* self, QPoint* offset) {
    auto* vqpicture = const_cast<VirtualQPicture*>(dynamic_cast<const VirtualQPicture*>(self));
    if (vqpicture && vqpicture->isVirtualQPicture) {
        return vqpicture->redirected(offset);
    } else {
        return ((VirtualQPicture*)self)->redirected(offset);
    }
}

// Base class handler implementation
QPaintDevice* QPicture_QBaseRedirected(const QPicture* self, QPoint* offset) {
    auto* vqpicture = const_cast<VirtualQPicture*>(dynamic_cast<const VirtualQPicture*>(self));
    if (vqpicture && vqpicture->isVirtualQPicture) {
        vqpicture->setQPicture_Redirected_IsBase(true);
        return vqpicture->redirected(offset);
    } else {
        return ((VirtualQPicture*)self)->redirected(offset);
    }
}

// Auxiliary method to allow providing re-implementation
void QPicture_OnRedirected(const QPicture* self, intptr_t slot) {
    auto* vqpicture = const_cast<VirtualQPicture*>(dynamic_cast<const VirtualQPicture*>(self));
    if (vqpicture && vqpicture->isVirtualQPicture) {
        vqpicture->setQPicture_Redirected_Callback(reinterpret_cast<VirtualQPicture::QPicture_Redirected_Callback>(slot));
    }
}

// Derived class handler implementation
QPainter* QPicture_SharedPainter(const QPicture* self) {
    auto* vqpicture = const_cast<VirtualQPicture*>(dynamic_cast<const VirtualQPicture*>(self));
    if (vqpicture && vqpicture->isVirtualQPicture) {
        return vqpicture->sharedPainter();
    } else {
        return ((VirtualQPicture*)self)->sharedPainter();
    }
}

// Base class handler implementation
QPainter* QPicture_QBaseSharedPainter(const QPicture* self) {
    auto* vqpicture = const_cast<VirtualQPicture*>(dynamic_cast<const VirtualQPicture*>(self));
    if (vqpicture && vqpicture->isVirtualQPicture) {
        vqpicture->setQPicture_SharedPainter_IsBase(true);
        return vqpicture->sharedPainter();
    } else {
        return ((VirtualQPicture*)self)->sharedPainter();
    }
}

// Auxiliary method to allow providing re-implementation
void QPicture_OnSharedPainter(const QPicture* self, intptr_t slot) {
    auto* vqpicture = const_cast<VirtualQPicture*>(dynamic_cast<const VirtualQPicture*>(self));
    if (vqpicture && vqpicture->isVirtualQPicture) {
        vqpicture->setQPicture_SharedPainter_Callback(reinterpret_cast<VirtualQPicture::QPicture_SharedPainter_Callback>(slot));
    }
}

// Derived class handler implementation
double QPicture_GetDecodedMetricF(const QPicture* self, int metricA, int metricB) {
    auto* vqpicture = const_cast<VirtualQPicture*>(dynamic_cast<const VirtualQPicture*>(self));
    if (vqpicture && vqpicture->isVirtualQPicture) {
        return vqpicture->getDecodedMetricF(static_cast<QPaintDevice::PaintDeviceMetric>(metricA), static_cast<QPaintDevice::PaintDeviceMetric>(metricB));
    } else {
        return ((VirtualQPicture*)self)->getDecodedMetricF(static_cast<QPaintDevice::PaintDeviceMetric>(metricA), static_cast<QPaintDevice::PaintDeviceMetric>(metricB));
    }
}

// Base class handler implementation
double QPicture_QBaseGetDecodedMetricF(const QPicture* self, int metricA, int metricB) {
    auto* vqpicture = const_cast<VirtualQPicture*>(dynamic_cast<const VirtualQPicture*>(self));
    if (vqpicture && vqpicture->isVirtualQPicture) {
        vqpicture->setQPicture_GetDecodedMetricF_IsBase(true);
        return vqpicture->getDecodedMetricF(static_cast<QPaintDevice::PaintDeviceMetric>(metricA), static_cast<QPaintDevice::PaintDeviceMetric>(metricB));
    } else {
        return ((VirtualQPicture*)self)->getDecodedMetricF(static_cast<QPaintDevice::PaintDeviceMetric>(metricA), static_cast<QPaintDevice::PaintDeviceMetric>(metricB));
    }
}

// Auxiliary method to allow providing re-implementation
void QPicture_OnGetDecodedMetricF(const QPicture* self, intptr_t slot) {
    auto* vqpicture = const_cast<VirtualQPicture*>(dynamic_cast<const VirtualQPicture*>(self));
    if (vqpicture && vqpicture->isVirtualQPicture) {
        vqpicture->setQPicture_GetDecodedMetricF_Callback(reinterpret_cast<VirtualQPicture::QPicture_GetDecodedMetricF_Callback>(slot));
    }
}

void QPicture_Delete(QPicture* self) {
    delete self;
}
